### ISSUE TITLE
We're looking for single file generators under new project system's project type instead of legacy

### ISSUE THREAD

**Issue by davkean** [dotnet/project-system#3535]
Single File Generators [are being looked up](https://docs.microsoft.com/en-us/visualstudio/extensibility/internals/registering-single-file-generators) under the new project types for C#, VB and F#. We should change this to look under the legacy project types so that providers don't need to register these in multiple places.

There's two places where we look up under the new GUIds:
- CPS\src\Microsoft.VisualStudio.ProjectSystem.VS.Implementation\VsProjectTypeValuesCacheService.cs
- https://github.com/dotnet/project-system/blob/b3b6ea07cd088926ac7973e1f916ec5be9df55e6/src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Generators/SingleFileGeneratorFactoryAggregator.cs#L63

This will be the reason no custom 3rd party single file generators work, because we don't find them.

Work to do:

- [x] Move VsSingleFileGeneratorFactory/SingleFileGeneratorFactoryAggregator -> CPS and change it call through CPS's VsProjectTypeValuesCacheService. This a project system feature and tied to the `SingleFileGenerators` capability and not to C#. [PR](https://devdiv.visualstudio.com/DevDiv/_git/CPS/pullrequest/122403?_a=files).
- [x] Implement the entire interface, not just the one method we have.
- [x] Delete our version [PR](https://github.com/dotnet/project-system/pull/3542)
- [x] Change VsProjectTypeValuesCacheService to pull ProjectTypeGuid from IItemTypeGuidProvider [PR](https://devdiv.visualstudio.com/DevDiv/_git/CPS/pullrequest/122410?_a=overview)
- [x] Remove our registration of single file generators from each package, these are the [RemoteCodeGeneratorRegistration] attributes. We don't need them after we start pulling from legacy. [PR](https://github.com/dotnet/project-system/pull/3543).
- [x] Null out fields in COM service

**Comment by davkean** [comment]
We should be pulling on the "active" IItemTypeGuidProvider.

**Comment by davkean** [comment]
We can delete our registration of them also from the C#/VB package.

**Comment by hallambaker** [comment]
Does not address my issue. I worked out that the GUID had changed and registered under the new ones. The problem is that the generators still don't run.

**Comment by davkean** [comment]
@hallambaker Can you paste your resulting pkgdef containing the registration?

**Comment by davkean** [comment]
Also can you tell me the exact way you are running the custom tool so that I can verify we're going down the same path?

**Comment by davkean** [comment]
I'm playing around with our Single File Generator registrations, such as ResX: https://github.com/dotnet/project-system/blob/master/src/Microsoft.VisualStudio.ProjectSystem.CSharp.VS/Packaging/CSharpProjectSystemPackage.cs#L28 and these run when I right-click on the RESX file and choose `Run Custom Tool`.

Looks like the platform doesn't log any error when the specified custom tool cannot be found or instantiated. I'll log a bug on that.

**Comment by hallambaker** [comment]
The code for all the tools is in this repo, there are about a dozen.
https://github.com/hallambaker/PHB-Build-Tools

One of the simplest tools is the one that generates the bindings to VSIX:
https://github.com/hallambaker/PHB-Build-Tools/tree/master/VSIXBuild/Goedel.Tool.VSIXBuild

The project that builds the tool set is here.
https://github.com/hallambaker/PHB-Build-Tools/tree/master/Pack/VSIX

These tools all run fine in a .net framework project.

**Comment by hallambaker** [comment]
OK, I think I might have it working now it is not clear exactly what has changed and I have been through the same exact steps with previous Visual Studio updates without success.

What I did was to create a completely new VSIX Project, then copied in the code for my generator from the original. It is now working correctly on a Framework and standard package in the debugger and after being installed.

I did have to add Microsoft.VisualStudio.Designer.Interfaces to the references but left the others alone. The main difference seems to be that several EnvDTE libraries have been added.

The same code does not work in the original project. It is possible that a different version of one of the libraries was being used that had not been updated to .NET standard.

It is not like I haven't tried this exact same approach before, I have. This time the problem seems to be fixed.

**Comment by hallambaker** [comment]
OK so now I have one working project and another not working in the exact same manner as before. Works as advertised for .net framework, fails for .net standard.

**Comment by hallambaker** [comment]
I can upload the VSIX packages if that helps.

**Comment by davkean** [comment]
Yes that would help, thanks.

**Comment by hallambaker** [comment]
This is the package that fails. It is based on a completely clean project that I just created. It works fine on a framework project and fails on a standard one.

[VSIX2.zip](https://github.com/dotnet/project-system/files/2020052/VSIX2.zip)

**Comment by hallambaker** [comment]
[Testsolution.zip](https://github.com/dotnet/project-system/files/2020054/Testsolution.zip)

These are the projects I tried them on. As you can see, works for Framework, fails for Standard.

**Comment by davkean** [comment]
I quickly looked at this and your registration for the new project type ({9A19103F-16F7-4668-BE54-9A1E7A4F7556}) looks incorrect;

You need to enclose the "project type" in braces:

```
[$RootKey$\Generators\{FAE04EC1-301F-11D3-BF4B-00C04F79EFBC}\Exceptional]
@="Exceptional"
"CLSID"="{7f47f99d-a484-4628-8ae4-0d866f7f7c43}"
"GeneratesDesignTimeSource"=dword:00000001
[$RootKey$\Generators\9A19103F-16F7-4668-BE54-9A1E7A4F7556\Exceptional] <-- Guid needs to be wrapped in braches
@="Exceptional"
"CLSID"="{7f47f99d-a484-4628-8ae4-0d866f7f7c43}"
"GeneratesDesignTimeSource"=dword:00000001
```

If I look at our code, similar to legacy we expect the braces to be present. Unlike legacy however, we don't complain when we can't find the generator.

**Comment by hallambaker** [comment]
Here is the weird part, I tried wrapping the GUID in braces in the declaration, the Registry keys are still bare! (And it still doesn't work).

These are the attributes I use to make the declarations:

   [ComVisible(true)]
    [Guid(GuidList.GuidExceptionalGeneratorString)]
    [ProvideObject(typeof(Exceptional))]
    [CodeGeneratorRegistration(typeof(Exceptional), "Exceptional", 
					"{FAE04EC1-301F-11D3-BF4B-00C04F79EFBC}", GeneratesDesignTimeSource = true)]
    [CodeGeneratorRegistration(typeof(Exceptional), "Exceptional", 
					"{9A19103F-16F7-4668-BE54-9A1E7A4F7556}", GeneratesDesignTimeSource = true)]
    [CodeGeneratorRegistration(typeof(Exceptional), "Exceptional", 
					"{694DD9B6-B865-4C5B-AD85-86356E9C88DC}", GeneratesDesignTimeSource = true)]

Both are wrapped the same way. But the keys generated are wrapped for the known generators and unwrapped for the others.

It looks to me like there is some logic stripping off the braces, but why? and how did my other one work?

I have tried a completely clean install of Visual Studio, no luck.

**Comment by hallambaker** [comment]
Here is the version with the braces in the declaration. 
[VSIX2 (2).zip](https://github.com/dotnet/project-system/files/2024783/VSIX2.2.zip)

Thanks for you help on this. I think there is something wrong somewhere though.

**Comment by davkean** [comment]
I installed the VSIX, and used `ASN2CS`, it was found all in types of C# projects:

```
Severity	Code	Description	Project	File	Line	Suppression State
Error		Parser Error Expected [Class] at Line 1 Col 6 in file Goedel.Registry.Source	NetStandard	
Warning		The custom tool 'ASN2CS' failed.  Parser Error Expected [Class] at Line 1 Col 6 in file Goedel.Registry.Source	NetFramework		
Error		Parser Error Expected [Class] at Line 1 Col 6 in file Goedel.Registry.Source	NetCore
```

**Comment by davkean** [comment]
It now looks like registration is found - let's make sure we're talking about the item types, what is "Build Action" set to when you are trying this?

**Comment by hallambaker** [comment]
I have tried all of them in turn. I mostly have it set to 'None' but I just tried C# compiler.

The registry key is still being set without braces. Changing the key name to add the braces does not seem to help.

**Comment by davkean** [comment]
The VSIX above (https://github.com/dotnet/project-system/files/2024783/VSIX2.2.zip) has the correct registration with the correct braces. Can you please try with that same build that you gave me?

**Comment by Dark-Bond** [comment]
Do you have an example of a complete project/package that I can build/install/run that will demonstrate a tool working from a .NET Standard 2.0 project inside Visual Studio 2017?  I see scraps of code referenced above, but I don't see a complete package.  If so, please publish the GitHub address.

**Comment by davkean** [comment]
Registration for single file generators for .NET Standard projects is identical to .NET Framework projects except instead of:

```
[CodeGeneratorRegistration(typeof(MyGenerator), "MyGenerator", "{FAE04EC1-301F-11D3-BF4B-00C04F79EFBC}", GeneratesDesignTimeSource = true)]
```
use:

```
[CodeGeneratorRegistration(typeof(MyGenerator), "MyGenerator", "{9A19103F-16F7-4668-BE54-9A1E7A4F7556}", GeneratesDesignTimeSource = true)]
```

Make note that the braces around the GUID _must be_ present.

**Comment by davkean** [comment]
Didn't mean to close this.

**Comment by bbowyersmyth** [comment]
We'll be looking to continue using RazorTemplatePreprocessor which we currently have in a PCL. I can't see how that is being registered https://github.com/mono/monodevelop/blob/master/main/src/addins/AspNet/Razor/Generator/RazorTemplatePreprocessor.cs
Will it work?

**Comment by davkean** [comment]
RazorTemplatePreprocessor won't work with .NET Standard/.NET Core projects unless someone uses the above registration, or until 15.8 Preview 3 when I've merged the fix in for this bug.

**Comment by hallambaker** [comment]
OK so got the chance to look at this again. I downloaded the same package you used from here just to be sure, deleted all my tools, and reinstalled from the package.

The bare registrations were still there but not wrapped. Searching for the string '9A19103F' in the registry produced no result. Looks to me like there is something up with the installer. Possibly those bare registrations are jamming it somehow. So I am now deleting the registry keys by hand and trying again. 

If this doesn't work, I think I will try uninstalling Visual Studio completely, reinstalling and trying again.

One thing I did notice is that all the generators are listed under version 14, not 15.

Computer\HKEY_CURRENT_USER\Software\Microsoft\VisualStudio\14.0_Config\Generators

**Comment by Pilchie** [comment]
In VS 2017, nothing will be directly in the registry - it will be in the private_registry.bin in the per-instance folder.

**Comment by hallambaker** [comment]
OK so the generators are now being registered correctly (hooray). That was probably happening all along and I was just seeing the old values. The keys under VS15 in the private hive are now correct.

Same behavior as before, the generators don't run.

Build action on the file is none, Custom Tool is CommandCS, exact same capitalization. Does not run in standard project, runs fine in framework.

**Comment by hallambaker** [comment]
OK, I think I might have the answer. While trying to narrow down the cause of the problem, I deleted all the references to the Mef components except for ASN2CS. It turns out that generator does work.

Tried recompiling with Mef references for all libraries that have project references. Did get a compile warning:

...\packages\Microsoft.VSSDK.BuildTools.15.6.152\tools\VSSDK\Microsoft.VsSDK.targets(82,5): warning VSSDK1018: Targets version (1.0.364.24297) and build task version (1.0.453.22492) do not match.  Restarting Visual Studio could help fix this issue.

...\packages\Microsoft.VSSDK.BuildTools.15.6.152\tools\VSSDK\Microsoft.VsSDK.targets(86,5): warning VSSDK1018: Targets version (15.6.152.24297) and build task version (15.7.104.22492) do not match.  Restarting Visual Studio could help fix this issue.

Tried it, and now the generators don't work in either standard or framework.

OK now trying with only the actual COM components listed, not the libraries they call.

**Comment by hallambaker** [comment]
And now the tools are not being registered at all!

I removed all the Mef references, didn't work, put in just the ASN2CS reference, put in all the references. None work.

I have pushed the sources of this build to the repository:
https://github.com/hallambaker/PHB-Build-Tools

This is the package
[VSIXtRY3.zip](https://github.com/dotnet/project-system/files/2040806/VSIXtRY3.zip)

**Comment by hallambaker** [comment]
OK so I re-installed the previous package (VSIX2). It did not work when I first installed the package. But I went out to dinner, came back and it is working.

So possibly there is some sort of cache issue involved. Perhaps it just takes a while for Visual Studio to process adding a dozen Domain Specific Languages at once. Or possibly there is some small change I made that is causing the install to fail.

Thanks for all the help!

**Comment by davkean** [comment]
After this change is merged in: https://github.com/dotnet/project-system/pull/3586 our behavior for where we look for IVsSingleFileGenerator & IVsEditorFactoryNotify implementations is identical to .NET Framework projects.

**Comment by hallambaker** [comment]
OK so it is mostly working now. One gotcha is that the matching of the custom tool is case insensitive for framework but case sensitive for standard.

So one other question is whether it will be possible to build VSIX plugins that make use of Standard libraries in the future. I don't much care about the VSIX package as that will only ever need to run on Windows. But I support the toolchain on OSX and Linux so right now I am using shared projects. If a VSIX plug in can call a standard library, I can eliminate the shared projects.

**Comment by davkean** [comment]
This hasn't actually been merged in on the platform side yet.

**Comment by Dark-Bond** [comment]
I can now confirm that the following creates a working Custom Tool in .NET Standard 2.0 libraries with Visual Studio 2017:

```
    [ComVisible(true)]
    [Guid("AFD180A2-A1F9-4429-9EE2-4118DD669960")]
    [CodeGeneratorRegistration(typeof(ResourceGenerator), nameof(ResourceGenerator), "{FAE04EC1-301F-11D3-BF4B-00C04F79EFBC}", GeneratesDesignTimeSource = true)]
    [CodeGeneratorRegistration(typeof(ResourceGenerator), nameof(ResourceGenerator), "{9A19103F-16F7-4668-BE54-9A1E7A4F7556}", GeneratesDesignTimeSource = true)]
    [CodeGeneratorRegistration(typeof(ResourceGenerator), nameof(ResourceGenerator), "{694DD9B6-B865-4C5B-AD85-86356E9C88DC}", GeneratesDesignTimeSource = true)]
    [ProvideObject(typeof(ResourceGenerator))]
    public sealed class ResourceGenerator : BaseCodeGeneratorWithSite
    {
    ....
    }

```

Thank you for your help.

**Comment by davkean** [comment]
This has been fixed and starting in 15.8 Preview 3 we'll look for single file generators in the same location as the legacy project system.
